kvbc/include/db_interfaces.h

```diff
@@ -35,13 +42,85 @@ class ILocalKeyValueStorageReadOnly {
 };
 
+// Categorized interfaces follow. Non-categorized interfaces will be deprecated after full integration of categorized
+// ones.
+
+// Add blocks to the key-value blockchain.
+class IBlockAdder {
+ public:
+  // Add a block from the given categorized updates and return its ID.
+  virtual BlockId add(categorization::Updates &&, const concordUtils::SpanWrapper &parent_span) = 0;
+
+  virtual ~IBlockAdder() = default;
+};
+
+// Key-value blockchain categorized read interface.
+//
+// Output vector parameters to multi* calls can contain more values than requested. This is an optimization, in order to
+// reduce memory allocations. Users are encouraged to reuse a single vector instance.
+class IReader {
+ public:
+  // Get the value of a key in `category_id` at `block_id`.
+  // Return std::nullopt if `key` doesn't exist at `block_id`.
+  virtual std::optional<categorization::Value> get(const std::string &category_id,
+                                                   const std::string &key,
+                                                   BlockId block_id) const = 0;
+
+  // Get the latest value of `key` in `category_id`.
+  // Return std::nullopt if the key doesn't exist or is deleted.
+  virtual std::optional<categorization::Value> getLatest(const std::string &category_id,
+                                                         const std::string &key) const = 0;
+
+  // Get values for keys at specific versions in `category_id`.
+  // `keys` and `versions` must be the same size.
+  // If a key is missing at the specified version or deleted, then std::nullopt is returned for it.
+  virtual void multiGet(const std::string &category_id,
+                        const std::vector<std::string> &keys,
+                        const std::vector<BlockId> &versions,
+                        std::vector<std::optional<Value>> &values) const = 0;
+
+  // Get the latest values of a list of keys in `category_id`.
+  // If a key is missing or is deleted, then std::nullopt is returned for it.
+  virtual void multiGetLatest(const std::string &category_id,
+                              const std::vector<std::string> &keys,
+                              std::vector<std::optional<Value>> &values) const = 0;
+
+  // Get the latest version of `key` in `category_id`.
+  // Return std::nullopt if the key doesn't exist or is deleted.
+  virtual std::optional<categorization::TaggedVersion> getLatestVersion(const std::string &category_id,
+                                                                        const std::string &key) const = 0;
+
+  // Get the latest versions of the given keys in `category_id`.
+  // If a key is missing, then std::nullopt is returned for its version.
+  virtual void multiGetLatestVersion(const std::string &category_id,
+                                     const std::vector<std::string> &keys,
+                                     std::vector<std::optional<categorization::TaggedVersion>> &versions) const;
+
```

Comment:
Looks good to me as far as we know where to take category_id from :-).

Reply:
It is application-specific and can be a mapping from key type to a category ID.

Reply:
@dartdart26 If `category_id` is application-specific then who is going to create those IDs and how are they generated?

Reply:
Yep :)

Reply:
Hi @dartdart26, can this one day be used, so that we store different types of Ethereum K/Vs in separate Merkle trees, each with its own root hash?

Reply:
@teoparvanov Yes, you can have multiple categories of the same type by just using different IDs.